PerfettoCds/Pipeline/PerfettoSourceParser.cs

```diff
@@ -0,0 +1,216 @@
+                            case PerfettoPluginConstants.SliceEvent:
+                                ev = new PerfettoSliceEvent();
+                                break;
+                            case PerfettoPluginConstants.ArgEvent:
+                                ev = new PerfettoArgEvent();
+                                break;
+                            case PerfettoPluginConstants.ThreadTrackEvent:
+                                ev = new PerfettoThreadTrackEvent();
+                                break;
+                            case PerfettoPluginConstants.ThreadEvent:
+                                ev = new PerfettoThreadEvent();
+                                break;
+                            case PerfettoPluginConstants.ProcessEvent:
+                                ev = new PerfettoProcessEvent();
+                                break;
+                            default:
+                                throw new Exception("Invalid event type");
+                        }
+                    }
+
+                    var colIndex = cellCount % numColumns;
+                    var colName = cols[colIndex].ToLower();
+
+                    // The event itself is responsible for figuring out how to process and store cell contents
+                    ev.ProcessCell(colName, cell, batch, stringCells, cellCounters);
+
+                    // If we've reached the end of a row, we've finished an event. Store it.
+                    if (++cellCount % numColumns == 0)
+                    {
+                        if (ev.GetType() == typeof(PerfettoSliceEvent))
+                        {
+                            // We get the timestamps used for displaying these events from the slice event
+                            if (traceStartTime == null)
+                            {
+                                traceStartTime = ((PerfettoSliceEvent)ev).Timestamp;
+                            }
+                            traceEndTime = ((PerfettoSliceEvent)ev).Timestamp;
+                        }
+
+                        // Store the event
+                        var result = dataProcessor.ProcessDataElement(ev, this, cancellationToken);
+
+                        ev = null;
+                    }
+                    IncreaseProgress(progressIncrease);
+                }
+            }
+
+            if (traceStartTime.HasValue)
+            {
+                // Use DateTime.Now as the wall clock time. This doesn't matter for displaying events on a relative timescale
+                // TODO Actual wall clock time needs to be gathered from SQL somehow
+                this.dataSourceInfo = new DataSourceInfo(traceStartTime.Value.ToNanoseconds, traceEndTime.Value.ToNanoseconds, DateTime.Now.ToUniversalTime());
+            }
+        }
+
+
+        public void ProcessSource(ISourceDataProcessor<PerfettoSqlEvent, PerfettoSourceParser, string> dataProcessor,
+            ILogger logger,
```

Comment:
I should have mentioned for the sql query diag logging we should have a timer to track how long the query took. Something like this (or wrap it in an auto logger class that upon destruction logs):
            var dateTimeQueryStarted = DateTime.UtcNow;
            // query here
            var dateTimeQueryFinished = DateTime.UtcNow;
            logger.Verbose($"Query completed in {(dateTimeQueryFinished - dateTimeQueryStarted).TotalSeconds}s at {dateTimeQueryFinished} UTC");

Reply:
Also add (to logging) from which table this is doing the selection....